### PR TITLE
GODRIVER-3049 Replace bsoncore.DocumentSequence with a bson type-agnostic analogue

### PR DESCRIPTION
**Pull request by prestonvasquez**
GODRIVER-3049

## Summary

The goal of bsoncore.Iterator is to abstract the arithmetic (i.e. constant time) retrieval of elements of a BSON array. The iterator does this by maintaining the “current” position of which element is “next-to-be-retrieved” in a type called List, which would likely be loaded from the cursor.nextBatch on a server response. This solution is more simple than it's predecessor and is more powerful, in that the List data is not restricted to documents.

Blank diagram (1)

## Background & Motivation

The DocumentSequence pattern was introduced in [GODRIVER-800](https://jira.mongodb.org/browse/GODRIVER-800) to avoid repetitively copying bytes into a document sequence. 

The 800 pattern assumes an enum-like type for categorizing data as either a sequence: i.e. {x:1}{x:2}{x:3}, or a BSON array: {1:{x:1},2:{x:2},3:{x:3}}. 3049 proposes we simplify this pattern using bsoncore.Array, which will handle both cases.

### CODE REVIEW

**Comment by blink1073** [comment]
@prestonvasquez this has a conflict now.

File: x/bsonx/bsoncore/iterator.go [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+import (
+	"errors"
+	"io"
+
+	"go.mongodb.org/mongo-driver/bson/bsontype"
+)
+
+// ErrCorruptedDocument is returned when a full document couldn't be read from
+// the sequence.
+var ErrCorruptedDocument = errors.New("invalid DocumentSequence: corrupted document")
```
Comment (matthewdale) [inline x/bsonx/bsoncore/iterator.go:18]: This error variable isn't used outside of this package and doesn't need to be exported.

File: x/bsonx/bsoncore/iterator.go [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+
+// ErrNonDocument is returned when a DocumentSequence contains a non-document
+// BSON value.
+var ErrNonDocument = errors.New("invalid DocumentSequence: a non-document value was found in sequence")
+
+// ErrInvalidDocumentSequenceStyle is returned when an unknown
+// DocumentSequenceStyle is set on a DocumentSequence.
+var ErrInvalidDocumentSequenceStyle = errors.New("invalid DocumentSequenceStyle")
```
Comment (matthewdale) [inline x/bsonx/bsoncore/iterator.go:26]: These error variables are unused and should be removed.

**Review by matthewdale — DISMISSED** [review]
Looks good! 👍

**Review by qingyang-hu — APPROVED** [review]
👍

**Review by qingyang-hu — APPROVED** [review]

**Review by matthewdale — APPROVED** [review]

**Pull request merged**